<a href="https://colab.research.google.com/github/Anavillag/Business-Analytics/blob/main/1_3_Reto_1_Ana_Mar%C3%ADa_Villa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 1 - Naive Bayes**

0. Cargar las librerias de trabajo

In [ ]:
import numpy as np  #Funciones matematicas
import pandas as pd #Conexiones conarchivos de excel

from sklearn.naive_bayes import GaussianNB #Modelo Gaussian Naive Bayes
from sklearn.metrics import confusion_matrix #Metricas- Como lo voy a evaluar


1. Se procede con la carga de los datos

In [ ]:
nxl='/content/1. BD2_CarCrash_Int.xlsx'
XDB=pd.read_excel(nxl, sheet_name=0)
XDB.dropna() #Elimina los valores nulos
XDB.head(100)

#Creamos el DataMart
XD=XDB[['Weekday','ViolCat','ClearWeather','Month','Daylight']] #Variables de Entrada
XD
yd=XDB[['Highway']] #Este es el valor deseado - que lo aprenda el modelo
yd

Highway
0           0
1           0
2           0
3           0
4           0
...       ...
6700        0
6701        0
6702        0
6703        0
6704        0

[6705 rows x 1 columns]

2. Se procede con la configuración y evaluación del modelo Naive Bayes (Autopista/Highways)


In [ ]:
from re import X
from functools import cmp_to_key
mnb=GaussianNB()
mnb.fit(XD,yd)

#Analizamos los resultados arrojados por el modelo
u=mnb.theta_  #Son las medias
print("Las medias de las variables son:\n", u)
sigma=np.sqrt(mnb.var_)
print("Los valores de las desviaciones son:\n",sigma)

#Se procede con el análisis de las variacioens de las variables
LS=u+sigma
print("Los limites superiores para las variables son:\n",LS)
LI=u-sigma
print("Los limites inferiores para las variables son:\n",LI)

#Se procede con la evaluación del comportamiento del modelo
#Prediciendo los valores de preaprobación del valor deseado
ydp=mnb.predict(XD) #Hizo el pronóstico de la Preaprobación para los solicitantes

cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es:\n",cm)
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]
print("VN=",VN)
print("FP=",FP)
print("FN=",FN)
print("VP=",VP)

#Exactitud
Ex=(VP+VN)/len(XD)
print("La exactitud del modelo es:\n",Ex)

#Tasa de Error
Ter=(FP+FN)/len(XD)
print("La tasa de error del modelo es:\n",Ter)

#Sensibilidad
Sen=VP/(VP+FN)
print("La sensibilidad del modelo es:\n",Sen)

#Especificidad
Esp=VN/(VN+FP)
print("La especificidad del modelo es:\n",Esp)

#Precisión
Pres=VP/(VP+FP)
print("La precisión del modelo es:\n",Pres)

#Predicción Negativo
PredNeg=VN/(VN+FN)
print("La predicción negativa del modelo es:\n",PredNeg)

#Predicción Positivo
PredPos=VP/(VP+FP)
print("La predicción positiva del modelo es:\n",PredPos)

#Para evlauar solicitante de credito
XSC=[[7,1,1,3,0]]
ydsc=mnb.predict(XSC)
print("La predicción del modelo es:\n",ydsc)


Las medias de las variables son:
 [[3.88945752 6.7778915  0.85261003 5.52548618 0.6814739 ]
 [3.82417582 4.31538462 0.79450549 5.47417582 0.70934066]]
Los valores de las desviaciones son:
 [[1.92831125 3.63475806 0.35449426 3.27379337 0.46590475]
 [1.82800803 2.36793707 0.40406253 3.26215279 0.45406663]]
Los limites superiores para las variables son:
 [[ 5.81776877 10.41264957  1.20710429  8.79927955  1.14737865]
 [ 5.65218385  6.68332168  1.19856802  8.73632861  1.16340729]]
Los limites inferiores para las variables son:
 [[1.96114628 3.14313344 0.49811577 2.25169281 0.21556915]
 [1.9961678  1.94744755 0.39044297 2.21202304 0.25527403]]
La matriz de confusión es:
 [[4560  325]
 [1563  257]]
VN= 4560
FP= 325
FN= 1563
VP= 257
La exactitud del modelo es:
 0.7184190902311708
La tasa de error del modelo es:
 0.2815809097688292
La sensibilidad del modelo es:
 0.1412087912087912
La especificidad del modelo es:
 0.9334698055271239
La precisión del modelo es:
 0.4415807560137457
La predicción 

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


3. Se procede con la estimación

In [ ]:
XDBn=pd.read_excel(nxl, sheet_name=0)
XDBn.dropna() #Para evitar celdas vacias
XDBn.head(100)

#Creamos el DataMart
XDn=XDBn [['Weekday', 'ViolCat','ClearWeather','Month','Daylight']] #Variables Entrada
ydn=XDBn[['Highway']] #Valor Deseado Aprenda el Modelo

mnb.fit(XDn,ydn)

#Hacer el pronostico para todos los dtaos disponibles
ydpn=mnb.predict(XDn)
print(ydpn)


#Se procede con la creación del informe
XDsal=np.column_stack((XDn,ydpn,ydn)) #Para mostrar columnas
df_sal=pd.DataFrame (XDsal) #Se crea una tabla
df_sal.to_excel("Informe-Reto1.xlsx")#Se envia al excel





/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


[0 0 0 ... 0 0 0]


**ANALISIS DE RESULTADOS**

Los resultados por el modelo propuesto cuenta con una base de datos que posee un total  de 6705 datos (accidentados en autopistas=582, accidentados no en autopistas=6123) , este mostró exactitud que alcanzó un porcentaje del 71,84% lo que indica que de un total de 100 datos de individuos, se presentan solo en 29 datos de inviduos se presenta un error a la hora de indicar si el accidente fue en la autopista o no. La métrica de sensibilidad alcanzó  un valor del  14,12% frente a la aprobación y una especificidad del 93,35%.Lo anterior muestra que el modelo funciona mejor para los casos en que el accidente no fue en un autopista a los que si. En adición,  la tasa de error del modelo es del 28,16%.
Por otro lado,  se evaluo una persona con las siguientes características: ['Weekday','ViolCat','ClearWeather','Month','Daylight']=[7,1,1,3,0] y con los datos arrojados se puede concluir que este no tendria un accidente automovilistico en una autopista. Finalmente, el modelo alcanzó una precisión del 44,16% y los accidentes en autopistas tiene un porcentaje en los datos del 28% y los accidentes no en autopistas de al rededor un 71%.